runtime/pkg/duckdbsql/duckdbsql.go

```diff
@@ -4,24 +4,163 @@ import (
 	"context"
 	databasesql "database/sql"
 	"database/sql/driver"
+	"encoding/json"
+	"regexp"
 	"sync"
 
 	"github.com/marcboeker/go-duckdb"
 )
 
+type AST struct {
+	sql       string
+	ast       astNode
+	rootNodes []*selectNode
+	aliases   map[string]bool
+	added     map[string]bool
+	fromNodes []*fromNode
+	columns   []*columnNode
+}
+
+type selectNode struct {
+	ast astNode
+}
+
+type columnNode struct {
+	ast astNode
+	ref *ColumnRef
+}
+
+type fromNode struct {
+	ast      astNode
+	parent   astNode
+	childKey string
+	ref      *TableRef
+}
+
+func Parse(sql string) (*AST, error) {
+	sqlAst, err := queryString("select json_serialize_sql(?::VARCHAR)", sql)
+	if err != nil {
+		return nil, err
+	}
+
+	nativeAst := astNode{}
+	err = json.Unmarshal(sqlAst, &nativeAst)
+	if err != nil {
+		return nil, err
+	}
+
+	ast := &AST{
+		sql:       sql,
+		ast:       nativeAst,
+		rootNodes: make([]*selectNode, 0),
+		aliases:   map[string]bool{},
+		added:     map[string]bool{},
+		fromNodes: make([]*fromNode, 0),
+		columns:   make([]*columnNode, 0),
+	}
+
+	err = ast.traverse()
+	if err != nil {
+		return nil, err
+	}
+	return ast, nil
+}
+
 // Format normalizes a DuckDB SQL statement
-func Format(sql string) (string, error) {
-	return queryString("SELECT json_deserialize_sql(json_serialize_sql(?::VARCHAR))", sql)
+func (a *AST) Format() (string, error) {
+	sql, err := json.Marshal(a.ast)
+	if err != nil {
+		return "", err
+	}
+	res, err := queryString("SELECT json_deserialize_sql(?::JSON)", string(sql))
+	return string(res), err
 }
 
-// Sanitize strips comments and normalizes a DuckDB SQL statement
-func Sanitize(sql string) (string, error) {
-	panic("not implemented")
+func (a *AST) GetTableRef() (*TableRef, bool) {
+	if len(a.fromNodes) == 1 {
+		return a.fromNodes[0].ref, true
+	}
+	return nil, false
+}
+
+// RewriteTableRefs replaces table references in a DuckDB SQL query
+func (a *AST) RewriteTableRefs(fn func(table *TableRef) (*TableRef, bool)) error {
+	for _, node := range a.fromNodes {
+		newRef, shouldReplace := fn(node.ref)
+		if !shouldReplace {
+			continue
+		}
+
+		if node.ref.Name == "" && newRef.Name != "" {
+			err := node.rewriteToBaseTable(newRef.Name)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
It looks like the node is not replaced in all cases.

If it's easiest to only support rewriting to base tables, that's okay, just put it in the docstring and return an error if attempting to rewrite to something else

Reply:
Also, will this preserve aliases in all cases? E.g. if rewriting `SELECT * FROM "path/to/file.csv" foo`, the output should be `SELECT * FROM new_table_name foo`

Reply:
Sounds good. I will remove the 1st check, we need to support base table to base table as well since plain urls like `from "s3:/path/to/data.csv"` needs to be replaced.

I have a test for making sure aliases are maintained. I get the alias from the existing node.